domain/application/state/application_endpoint.go

```diff
@@ -74,21 +74,57 @@ func (st *State) GetApplicationEndpointNames(ctx context.Context, appUUID coreap
 	return transform.Slice(eps, func(r charmRelationName) string { return r.Name }), nil
 }
 
+// ValidateEndpointBindingsForApplication validates the provided endpoint bindings
+// can be applied to the specified application.
+// This checks that:
+//   - each of the machines this application is deployed to has an address in the
+//     target spaces.
+//
+// TODO(jack-w-shaw): Implement this method. Look for the `validateForMachines()`
+// method in state/endpoint_bindings.go on 3.x branch(es).
```

Comment:
This should be done in this PR or a subsequent PR, otherwise remove the TODO and add a jira card.